### PR TITLE
Add HTTP::Request#route_uri_pattern that returns URI pattern of matched route.

### PR DESCRIPTION
**Pull request by joelhawksley**
### Motivation / Background

In the GitHub RoR monolith, we output the route URI pattern in an HTML meta tag in our application layout for analysis purposes. However, our current implementation is quite manual.

### Detail

This change adds an attribute to requests with the URI pattern of the matched route.

### Additional information

Co-authored-by: Kate Higa <khiga8@github.com>

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: actionpack/lib/action_dispatch/http/request.rb [hunk 1]
```diff
@@ -57,6 +57,8 @@ def self.empty
       new({})
     end
 
+    attr_accessor :route_uri_pattern
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/http/request.rb:60]: We should implement this using the rack env so middleware can also have access to the information. See `routes=` for example.

**Comment by joelhawksley** [comment]
@rafaelfranca thanks for the suggestions in your review. I've updated the PR accordingly ❤️

File: actionpack/lib/action_dispatch/http/request.rb [hunk 2]
```diff
@@ -153,6 +153,14 @@ def routes=(routes) # :nodoc:
       set_header("action_dispatch.routes", routes)
     end
 
+    def route_uri_pattern # :nodoc:
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/http/request.rb:156]:
```suggestion
    def route_uri_pattern
```

We want this method to be public API so we need to document it.

File: actionpack/lib/action_dispatch/http/request.rb [hunk 3]
```diff
@@ -145,6 +145,18 @@ def request_method
...
 
+    # Returns the URI pattern of the matched route for the request,
+    # using the same format as `bin/rails routes`:
+    #
+    #   request.route_uri_pattern # => "/:controller(/:action(/:id))(.:format)"
+    def route_uri_pattern
+      get_header("action_dispatch.route_uri_pattern")
+    end
+
+    def route_uri_pattern=(pattern) # :nodoc:
```
Comment (zzak) [inline actionpack/lib/action_dispatch/http/request.rb:156]: @joelhawksley I've opened #47306 but I was thinking we might want to add a test for this if possible? WDYT?

**Pull request merged**